### PR TITLE
Add Apache Doris SQL dialect support to SQLFluff

### LINKED ISSUE #6978: Apache Doris Dialect Support
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

Support SQL Dialect for [Apache Doris](https://doris.apache.org/)

### Use case

_No response_

### Dialect

Apache Doris

### Are you willing to work on and submit a PR to address the issue?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by morningman**
Fix #6978

### Description

This PR adds comprehensive support for the Apache Doris SQL dialect to SQLFluff, extending the existing MySQL grammar with Doris-specific syntax features. The implementation includes support for CREATE TABLE, DROP TABLE, INSERT statements, and various Doris-specific table properties and clauses.

#### 1. **CREATE TABLE Statement Support**
- **Engine Types**: Support for Doris-specific engines including `olap`, `mysql`, `elasticsearch`, `hive`, `hudi`, `iceberg`, `jdbc`, `broker`
- **Key Types**: Support for `DUPLICATE KEY`, `AGGREGATE KEY`, and `UNIQUE KEY` syntax
- **Partitioning**: 
  - Auto partitioning with `AUTO PARTITION BY RANGE(function) ()`
  - Manual range partitioning with `PARTITION BY RANGE(columns)`
  - List partitioning with `PARTITION BY LIST(columns)`
- **Distribution**: Support for `DISTRIBUTED BY HASH(columns)` and `DISTRIBUTED BY RANDOM`
- **Rollup Definitions**: Support for `ROLLUP` clauses
- **Table Properties**: Support for `PROPERTIES` clauses with key-value pairs
- **Index Definitions**: Support for Doris-specific index types (`INVERTED`, `BITMAP`, `BLOOM_FILTER`)
- **Generated Columns**: Support for `GENERATED ALWAYS AS` syntax
- **CREATE TABLE ... AS SELECT (CTAS)**: Full support for CTAS with optional properties
- **CREATE TABLE ... LIKE**: Support for creating tables based on existing table structure
- **External Tables**: Support for `CREATE EXTERNAL TABLE` and `CREATE TEMPORARY EXTERNAL TABLE`

#### 2. **DROP TABLE Statement Support**
- Support for `DROP TABLE [IF EXISTS] [db_name.]table_name [FORCE]` syntax
- Database-qualified table names
- Optional `IF EXISTS` clause
- Optional `FORCE` keyword for unrecoverable table deletion

#### 3. **INSERT Statement Support**
- Basic `INSERT INTO table VALUES (...)` syntax
- Column specification with `INSERT INTO table (col1, col2) VALUES (...)`
- `DEFAULT` value support in VALUES clauses
- Multiple row insertion with comma-separated VALUES
- `INSERT ... SELECT` statements
- **Partition specification** with `PARTITION (p1, p2)` clause
- **Label specification** with `WITH LABEL label1` clause
- Complex combinations of all INSERT features

#### 4. **Doris-Specific Keywords and Grammar**
- Added Doris-specific reserved and unreserved keywords
- Support for Doris aggregation functions (`MAX`, `MIN`, `REPLACE`, `SUM`, `BITMAP_UNION`, `HLL_UNION`, `QUANTILE_UNION`)
- Support for `STORED` and `VIRTUAL` keywords in generated columns
- Proper handling of Doris-specific data types and constraints

### Technical Implementation

- **Base Dialect**: Extends MySQL dialect to leverage existing MySQL grammar compatibility
- **Custom Segments**: Implements Doris-specific grammar segments for complex syntax
- **Keyword Management**: Properly manages Doris-specific keywords without conflicts
- **Grammar Extensions**: Adds Doris-specific grammar rules while maintaining compatibility

### Testing

Comprehensive test coverage includes:
- **CREATE TABLE tests**: 15+ test files covering various table creation scenarios
- **DROP TABLE tests**: 4 test files covering different drop scenarios
- **INSERT tests**: 9 test files covering various insert patterns
- **Hive integration tests**: Multiple test files for Hive catalog integration
- **Complex syntax tests**: Edge cases and combinations of multiple features

### Compatibility

- **MySQL Compatibility**: Maintains compatibility with existing MySQL syntax
- **Doris Standards**: Follows official Apache Doris documentation and syntax
- **Backward Compatibility**: No breaking changes to existing functionality

### Documentation

The implementation follows the official Apache Doris documentation:
- [CREATE TABLE](https://doris.apache.org/docs/dev/sql-manual/sql-statements/table-and-view/table/CREATE-TABLE)
- [DROP TABLE](https://doris.apache.org/docs/dev/sql-manual/sql-statements/table-and-view/table/DROP-TABLE)
- [INSERT](https://doris.apache.org/docs/dev/sql-manual/sql-statements/data-modification/DML/INSERT)

### Files Changed

- `src/sqlfluff/dialects/dialect_doris.py` - Main dialect implementation
- `test/fixtures/dialects/doris/` - Comprehensive test suite (30+ test files)

### Pull Request checklist

- [ ] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_doris.py [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+class StatementSegment(mysql.StatementSegment):
+    """Overriding StatementSegment to allow for additional segment parsing."""
+
+    match_grammar = mysql.StatementSegment.match_grammar.copy(
+        insert=[
+            # Add any Doris-specific statement types here
+            Ref("DropTableStatementSegment"),
+            Ref("InsertStatementSegment"),
+        ]
+    )
```
Comment (keraion) [inline src/sqlfluff/dialects/dialect_doris.py]: nit: newline

File: test/fixtures/dialects/doris/.sqlfluff [hunk 2]
```diff
@@ -0,0 +1,2 @@
+[sqlfluff]
+dialect = doris
```
Comment (keraion) [inline test/fixtures/dialects/doris/.sqlfluff]: nit: newline

File: test/fixtures/dialects/doris/create_hive_table_as_select_basic.sql [hunk 3]
```diff
@@ -0,0 +1,3 @@
+CREATE TABLE hive_catalog.hive_db.hive_table
+ENGINE=hive
+AS SELECT * FROM source_table;
```
Comment (keraion) [inline test/fixtures/dialects/doris/create_hive_table_as_select_basic.sql]: nit: newline

File: test/fixtures/dialects/doris/create_hive_table_as_select_partitioned.sql [hunk 4]
```diff
@@ -0,0 +1 @@
+
```
Comment (keraion) [inline test/fixtures/dialects/doris/create_hive_table_as_select_partitioned.sql]: This appears to be an empty file.

File: test/fixtures/dialects/doris/create_hive_table_as_select_with_comment.sql [hunk 5]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE hive_catalog.hive_db.hive_table_with_comment
+ENGINE=hive
+COMMENT 'This is a Hive table created as select.'
+PROPERTIES (
+  'file_format' = 'parquet'
+)
+AS SELECT id, name FROM source_table;
```
Comment (keraion) [inline test/fixtures/dialects/doris/create_hive_table_as_select_with_comment.sql]: nit: newline

File: test/fixtures/dialects/doris/create_hive_table_as_select_with_properties.sql [hunk 6]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE hive_catalog.hive_db.hive_table
+ENGINE=hive
+PROPERTIES (
+  'file_format' = 'parquet',
+  'hive.metastore.uris' = 'thrift://127.0.0.1:9083'
+)
+AS SELECT * FROM source_table;
```
Comment (keraion) [inline test/fixtures/dialects/doris/create_hive_table_as_select_with_properties.sql]: nit: newline

File: test/fixtures/dialects/doris/create_hive_table_basic.sql [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+(
+    id INT,
+    name STRING,
+    age INT,
+    email STRING
+)
+ENGINE=hive
+PROPERTIES (
+    'file_format' = 'parquet'
+);
```
Comment (keraion) [inline test/fixtures/dialects/doris/create_hive_table_basic.sql]: nit: newline

File: src/sqlfluff/dialects/dialect_doris.py [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+            "MAX",
+            "MIN", 
+            "REPLACE",
+            "SUM",
+            "BITMAP_UNION",
+            "HLL_UNION",
+            "QUANTILE_UNION",
+            optional=True,
+        ),
+    )
```
Comment (keraion) [inline src/sqlfluff/dialects/dialect_doris.py]: This should be able to be simplified with a `copy`
```suggestion
class ColumnDefinitionSegment(mysql.ColumnDefinitionSegment):
    """A column definition, e.g. for CREATE TABLE or ALTER TABLE.

    Doris-specific version that supports aggregation functions like MAX, MIN, REPLACE, SUM.
    """

    match_grammar = mysql.ColumnDefinitionSegment.match_grammar.copy(
        insert=[
            OneOf(
                "MAX",
                "MIN",
                "REPLACE",
                "SUM",
                "BITMAP_UNION",
                "HLL_UNION",
                "QUANTILE_UNION",
                optional=True,
            ),
        ]
    )
```

File: src/sqlfluff/dialects/dialect_doris.py (hunk 1 above)
Comment (keraion) [inline src/sqlfluff/dialects/dialect_doris.py]: These two segments appear to already exist from `ansi` so this shouldn't be required.

**Comment by morningman** [comment]
Hi @keraion , thanks for your review. I fixed all issues. PTAL

**Comment by keraion** [comment]
Would you mind running the pre-commit hooks on these files? If using tox `tox -e pre-commit -- run --all` or without `pre-commit run --all`

**Comment by morningman** [comment]
> Would you mind running the pre-commit hooks on these files? If using tox `tox -e pre-commit -- run --all` or without `pre-commit run --all`

So I need to run this command in my local env? What is it about to do?

**Comment by keraion** [comment]
Yes, this will handle a few of the linting and formatting issues that remain. We have guides on [getting tox setup](https://github.com/sqlfluff/sqlfluff/blob/main/CONTRIBUTING.md#requirements), [how to run the hooks automatically when committing](https://github.com/sqlfluff/sqlfluff/blob/main/CONTRIBUTING.md#pre-commit-config), and a [last checks guide](https://github.com/sqlfluff/sqlfluff/wiki/Contributing-Dialect-Changes#final-checks-before-committing) that very briefly goes over it, but this command will run all the pre-commit hooks that we have in the `.pre-commit-config.yaml`.

Namely this will run black, ruff, and a few other helpers to clean up. These should be able to autofix _most_ issues, but there may be a few items that need to be manually addressed.

LMK if you have any other questions!

**Comment by morningman** [comment]
Hi @keraion , I ran the tox and reformat the code, now it looks good:
```
tox -e pre-commit -- run --all
pre-commit: commands[0]> pre-commit run --all
don't commit to branch...................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
black....................................................................Passed
mypy.....................................................................Passed
flake8...................................................................Passed
doc8.....................................................................Passed
yamllint.................................................................Passed
ruff.....................................................................Passed
codespell................................................................Passed
  pre-commit: OK (131.68=setup[0.10]+cmd[131.57] seconds)
  congratulations :) (132.00 seconds)
```

**Review by keraion — APPROVED** [review]
Awesome work! Thanks so much for the contribution. 🎉
